target: add support for Loongson LoongArch64 (kernel 6.6)

hackpascal opened the pull request:
Add target for Loongson LoongArch64-based boards.

LoongArch is a new RISC ISA developed by Loongson. It's a bit like MIPS or RISC-V.
LoongArch includes both 32-bit and 64-bit versions (LoongArch32/LoongArch64).

Loongson 3A5000 and 3A6000 are the two existing CPUs of LoongArch64 and is used for PC products.
It's BIOS supports ACPI and UEFI-only boot. These CPUs supports SMP and SMT.

At present only LoongArch64 is supported by linux kernel.

Toolchain requirement:
binutils >= 2.40
gcc >= 13.1
glibc >= 2.36
musl >= 1.2.5

Installation:
Write openwrt-loongarch64-generic-ext4-combined-efi.img directly to a hard disk.

Note:
mac80211 type drivers are now available.
Both integrated graphics (7A2000) and discrete graphics (amdgpu only) works.
Serial port ttyS0 is also enabled for kernel console.
Tested on Loongson-3A6000-HV-7A2000-XA61200.

For details, please check the following links:
https://lwn.net/Articles/861951/
https://loongson.github.io/LoongArch-Documentation/README-EN.html

The picture of XA61200 motherboard is as follows:
https://github.com/loongson/Firmware/blob/main/Image/XA61200.jpg
![image](https://raw.githubusercontent.com/loongson/Firmware/main/Image/XA61200.jpg)

hackpascal committed:
loongarch64: new target

Add target for Loongson LoongArch64-based boards.

LoongArch is a new RISC ISA developed by Loongson. It's a bit like
MIPS or RISC-V. LoongArch includes both 32-bit and 64-bit versions
(LoongArch32/LoongArch64).

Loongson 3A5000 and 3A6000 are the two existing CPUs of LoongArch64
and is used for PC products. It's BIOS supports ACPI and UEFI-only
boot. These CPUs supports SMP and SMT.

At present only LoongArch64 is supported by linux kernel.

Toolchain requirement:
binutils >= 2.40
gcc >= 13.1

For details, please check the following links:
https://lwn.net/Articles/861951/
https://loongson.github.io/LoongArch-Documentation/README-EN.html

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

hackpascal committed:
kernel: modules: netdevices: add depedency required for loongarch64

Add TARGET_loongarch64 as dependency for kmod-mdio-devres,
kmod-mdio-gpio and kmod-switch-rtl8366-smi

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

hackpascal committed:
kernel: modules: video: adapt for loongarch64

* Allow kmod-acpi-video to be built for loongarch64:
The x86-specific CONFIG_ACPI_WMI will be split from default
kmod-acpi-video as a board-specific addition.

* Allow kmod-drm-amdgpu to be built for loongarch64:
Also add loongarch64-specific configs and modules.

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

hackpascal committed:
grub2: add EFI support for loongarch64

Add a new package for loongarch64 which only supports EFI.

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

hackpascal committed:
toolchain: Disable libtsan and liblsan sanitizer on loongarch64

libtsan and liblsan are not supported by glibc on loongarch64

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

cherinyy commented:
Please refrain from sending meaningless comments and focus on development communication.

castillofrancodamian commented:
6.6 is [LTS](https://kernel.org/category/releases.html).

cherinyy commented:
> 6.6 is [LTS](https://kernel.org/category/releases.html).

yes, it is. but right now 6.6 is still stable because 6.7 is still mainline.

the longterm maintenance version of the 6.6 kernel has still not been released.

hackpascal commented:
https://github.com/openwrt/openwrt/actions/runs/7448683628/job/20263661083?pr=14357

> ====== Make errors from logs/target/linux/compile.txt ======
  HDRINST usr/include/asm/ioctls.h
  HDRINST usr/include/asm/stat.h
  HDRINST usr/include/asm/bitsperlong.h
  INSTALL /__w/openwrt/openwrt/openwrt/build_dir/target-loongarch64_generic_glibc/linux-loongarch64/linux-6.6.9/user_headers/include
make[4]: Leaving directory '/__w/openwrt/openwrt/openwrt/build_dir/target-loongarch64_generic_glibc/linux-loongarch64/linux-6.6.9'
make[6]: hexdump: No such file or directory
make[6]: hexdump: No such file or directory
make[6]: hexdump: No such file or directory
make[6]: hexdump: No such file or directory
truncate: Invalid number: 'arch/loongarch/boot/vmlinux.bin'
make[6]: *** [drivers/firmware/efi/libstub/Makefile.zboot:13: arch/loongarch/boot/vmlinux.bin] Error 1

This is weird. Can hexdump be missing on the build system?

hackpascal committed:
openssl: add linux64-loongarch64 into the targets list

Add "linux64-loongarch64-openwrt" into openssl configurations to enable
building on loongarch64 machines.

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

ynezz commented:
> Can hexdump be missing on the build system?

Indeed, its not present, because its not current build host dependency defined and assured in `include/prereq-build.mk`. You can check it yourself in the CI tools container:

```
$ docker run --rm -it ghcr.io/openwrt/tools:latest which hexdump
```

[Containers used on CI](https://github.com/openwrt/actions-shared-workflows/tree/main/.github/dockerfiles) are basically reusing buildbot's [buildworker container](https://github.com/openwrt/buildbot/blob/master/docker/buildworker/Dockerfile) used on https://buildbot.openwrt.org

Anyways, that new `hexdump` dependency seems to be introduced in 6.4 via commit [bca2f3a9406b](https://github.com/torvalds/linux/commit/bca2f3a9406b) `("efi/zboot: Add BSS padding before compression")` "just" for some padding purposes, so either we debloat it upstream by some other means or just bite the bullet and introduce `hexdump` as a new host build dependency _OR_ as a new compiled host tool under `tools`.

hackpascal commented:
> a new compiled host tool under `tools`

I‘ll give this a try.

hackpascal commented:
At least all checks passed after introducing util-linux and disabling mac80211 drivers.

KFERMercer commented:
build failed: libffi requires at least version [v3.4.3](https://github.com/libffi/libffi/tree/v3.4.3), now is v3.4.2.

hackpascal commented:
> build failed: [libffi](https://github.com/openwrt/packages/blob/master/libs/libffi/Makefile) requires at least [v3.4.3](https://github.com/libffi/libffi/releases/tag/v3.4.3), now is v3.4.2.

@KFERMercer libffi is not a base package included in the openwrt repo.
It should be updated in the packages repo by another PR.

KFERMercer commented:
> @KFERMercer libffi is not a base package included in the openwrt repo. It should be updated in the packages repo by another PR.

sorry, just mention for other people

cherinyy commented:
Kernel 6.6.14 is now available and it is the longterm version.

hackpascal commented:
> Kernel 6.6.14 is now available and it is the longterm version.

OK. I'll update later

dangowrt commented:
@hackpascal Skipping Linux 6.1 and preparing the next release based on Linux 6.6 is worth the debate -- the number of backport patches on top of 6.1 is already nearly unbearable by now and using Linux 6.6 as a base will allow us to drop most of those obviously and rather make a release happen sooner than later. Maybe you should open another PR only covering the part introducing support for Linux 6.6?

On a personal note, switching to Linux 6.6 has a sad side to me because it drops the support for ox820 SoC (2x ARM11MPCore, 2x SATA, 1x gigE, PCIe1 x1, USB 2.0 x2, parallel NAND to boot, SPARC LEON uC remoteproc :rocket: , even a weird 2D graphics unit) which I had worked on a lot in my free time many years ago. And switching to Linux 6.6 implies dropping the oxnas target in OpenWrt :cry:

hackpascal commented:
> Maybe you should open another PR only covering the part introducing support for Linux 6.6?

It's reasonable. I'll do this after Chinese New Year.

t-8ch commented on tools/Makefile:
```diff
@@ -128,6 +129,7 @@ $(curdir)/quilt/compile := $(curdir)/autoconf/compile $(curdir)/findutils/compil
 $(curdir)/sdcc/compile := $(curdir)/bison/compile
 $(curdir)/squashfs3-lzma/compile := $(curdir)/lzma-old/compile
 $(curdir)/squashfs4/compile := $(curdir)/xz/compile $(curdir)/zlib/compile
+$(curdir)/util-linux/compile := $(curdir)/meson/compile $(curdir)/bison/compile
```
As you are using the autotools build system the dependency on meson should be unnecessary.

hackpascal replied:
Yep. I was trying meson first but it can't precisely select the minimum set of programs to build. That's why now autotools build system is used.
I'll remove this line.

oliv3r commented:
@hackpascal what's the status in having a seperate branch for just the 6.6 bump related changes?

hackpascal commented:
> @hackpascal what's the status in having a seperate branch for just the 6.6 bump related changes?

should be started in next week lol.

oliv3r commented:
> > @hackpascal what's the status in having a seperate branch for just the 6.6 bump related changes?
> 
> should be started in next week lol.

Perfect, then I won't pull your branch to base my changes upon and wait for your 6.6 branch :) Feel free to ping me on the new MR :)

hackpascal commented:
@oliv3r @dangowrt 
I have rebased this PR on the newest main branch.
The linux-6.6 only PR is also created: #14751

wigyori commented:
@hackpascal - do you reckon if the LS2K0500 could be supported within the same target? As I understand that's also a 64-bit LoongArch-based CPU (LA264-core).

hackpascal commented:
> @hackpascal - do you reckon if the LS2K0500 could be supported within the same target? As I understand that's also a 64-bit LoongArch-based CPU (LA264-core).

Should be supported, and maybe with some config changes. But I don't have such board to test now.

wigyori commented:
> > @hackpascal - do you reckon if the LS2K0500 could be supported within the same target? As I understand that's also a 64-bit LoongArch-based CPU (LA264-core).
> 
> Should be supported, and maybe with some config changes. But I don't have such board to test now.

I do have, so I'll get to work on that. :)

hackpascal committed:
libunwind: add support for loongarch64

Modify package depends to allow building for loongarch64.
Also fix for building with musl.

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

hauke commented:
Please rebase this PR. The musl libc, elfutils and GCC changes are already merged upstream and not needed here any more.

hackpascal commented:
> Please rebase this PR. The musl libc, elfutils and GCC changes are already merged upstream and not needed here any more.

@hauke Ah. finally.
Thanks for reminding

yichya commented:
> Kernel can't recognize /sbin/init as an executable program. This issue will take time to be analyzed.
> 
> So.. I still have to disable musl for loongarch64 for now. At lease glibc does work well.

This is caused by incorrect musl symlink filename (it is created as `ld-musl-loongarch64.so.1` while it should be `ld-musl-loongarch-lp64d.so.1`). 



Create a symlink there (I did a quick hack by creating it in `package/base-files`) and the problem will be solved, however a more proper fix is still needed to be figured out.

![image](https://github.com/openwrt/openwrt/assets/3683052/f89cb639-e5c0-489c-bbbc-c18a648f0c7a)

hackpascal commented:
> * Squashfs images does not work because 16KiB page size is not supported by f2fs in Linux 6.6.

That's why I disabled squashfs for now lol

hackpascal committed:
toolchain/musl: fix loongarch64 ldso file name

GCC has changed musl dynamic linker name from
ld-musl-loongarch-lp64d.so.1 to ld-musl-loongarch64.so.1 recently [1].

This means there are two dynamic linker names will be used across different
ersions of GCC. But musl 1.2.5 only supports the new name while the GCC
we're currently using uses the old name.

To maintain compatibility with all versions of GCC, the musl is then patched
to generate two symbolic links to libc.so with both old and new names.

[1] https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=8bccee51f0deac64b79cd9ad75df599422f4c8ff

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

hackpascal commented:
musl now fixed :smiley:

hackpascal commented:
> Squashfs images does not work because 16KiB page size is not supported by f2fs in Linux 6.6.
> 
> * ![image](https://private-user-images.githubusercontent.com/3683052/323125136-46e05ce2-4298-414e-a736-a5c0205ac44e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzNjk2ODIsIm5iZiI6MTcxMzM2OTM4MiwicGF0aCI6Ii8zNjgzMDUyLzMyMzEyNTEzNi00NmUwNWNlMi00Mjk4LTQxNGUtYTczNi1hNWMwMjA1YWM0NGUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MTdUMTU1NjIyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDAxNGRmNTA2NTMwMzFlNDQxMTYwMGE4MTNmYmVjOTY5ZjYzZmEwZDk4Mjc4OTQ3YmQwM2JjODg3MTEyMDNiNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.zgDysFOhfugtgMKHe0TXSqmkYwYwnhhYFsEFhKNJOwg)
> * Maybe try [torvalds/linux@aea6bf9](https://github.com/torvalds/linux/commit/aea6bf908d730b01bd264a8821159db9463c111c) or just use another overlay filesystem like ext4

It's better to do this in another PR

hackpascal committed:
toolchain/gcc: fix loongarch64 ldso file name

GCC has changed musl dynamic linker name from
ld-musl-loongarch-lp64d.so.1 to ld-musl-loongarch64.so.1 recently [1].

Meanwhile musl 1.2.5 only supports the new name. So it's better to follow
the new name.

[1] https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=8bccee51f0deac64b79cd9ad75df599422f4c8ff

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

981213 commented:
Merged. Thanks!
